ledger-core/insolar/payload/validation_vdelegatedrequestfinished.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 Insolar Network Ltd.
+// All rights reserved.
+// This material is licensed under the Insolar License version 1.0,
+// available at https://github.com/insolar/assured-ledger/blob/master/LICENSE.md.
+
+package payload
+
+import (
+	"github.com/insolar/assured-ledger/ledger-core/insolar/contract"
+	"github.com/insolar/assured-ledger/ledger-core/pulse"
+	"github.com/insolar/assured-ledger/ledger-core/vanilla/throw"
+)
+
+var _ Validatable = &VDelegatedRequestFinished{}
+
+func (m *VDelegatedRequestFinished) validateUnimplemented() error {
+	switch {
+	case m.ResultFlags != nil:
+		return throw.New("ResultFlags should be empty")
+	case m.EntryHeadHash != nil:
+		return throw.New("EntryHeadHash should be empty")
+	}
+	return nil
+}
+
+func (m *VDelegatedRequestFinished) isIntolerable() bool {
+	return m.GetCallFlags().GetInterference() == contract.CallIntolerable
+}
+
+func (m *VDelegatedRequestFinished) Validate(currentPulse pulse.Number) error {
+	if err := m.validateUnimplemented(); err != nil {
+		return err
+	}
+
+	switch m.GetCallType() {
+	case CTMethod, CTConstructor:
+	case CTInboundAPICall, CTOutboundAPICall, CTNotifyCall, CTSAGACall, CTParallelCall, CTScheduleCall:
+		return throw.New("CallType is not implemented")
+	case CTInvalid:
+		fallthrough
+	default:
+		return throw.New("CallType must be valid")
+	}
+
+	callFlags := m.GetCallFlags()
+	if f := callFlags.GetState(); f >= contract.StateFlagCount || f <= contract.StateInvalid {
+		return throw.New("CallFlags state should be valid")
+	} else if f := callFlags.GetInterference(); f >= contract.InterferenceFlagCount || f <= contract.InterferenceInvalid {
+		return throw.New("CallFlags interference should be valid")
+	}
+
+	if !m.Callee.IsSelfScope() {
+		return throw.New("Callee must be valid self scoped Reference")
+	} else if pn := m.Callee.GetLocal().Pulse(); !pn.IsTimePulse() || pn >= currentPulse {
+		return throw.New("Callee must have valid time pulse with pulse lesser than current pulse")
+	}
+
+	if m.CallIncoming.IsEmpty() {
+		return throw.New("CallIncoming should be non-empty")
+	} else if pn := m.CallIncoming.GetLocal().Pulse(); !pn.IsTimePulse() || pn >= currentPulse {
```

Comment:
pulse.Number has nice readable helpers:
```
IsBefore(pn Number) bool
IsAfter(pn Number) bool
IsBeforeOrEq(pn Number) bool
IsEqOrAfter(pn Number) bool
```
I think it would be easier to read these conditions with helpers

Reply:
`m.CallIncoming.GetBase().Pulse() > m.CallIncoming.GetLocal().Pulse()`
Do you think that's harder to read than
`m.CallIncoming.GetBase().Pulse().IsAfter(m.CallIncoming.GetLocal().Pulse())`
?

Reply:
`pn >= currentPulse` is easier, probably, than `pn.IsAfterOrEq(currentPulse)`, but that will lead to inconsistency in that two cases.

Reply:
if you put those function callse to variables
```
incomingLocalPulse.IsAfterOrEq(currentPulse)

incomingBasePulse.IsAfter(incomingLocalPulse)
```
then this will be read like a sentence

Reply:
fixed